static/app/utils/featureObserver.ts

```diff
@@ -1,71 +1,109 @@
 import type {Flags} from 'sentry/types/event';
 import type {Organization} from 'sentry/types/organization';
+import type {Project} from 'sentry/types/project';
 
+export const FEATURE_FLAG_BUFFER_SIZE = 100;
 let __SINGLETON: FeatureObserver | null = null;
 
 export default class FeatureObserver {
   /**
    * Return the same instance of FeatureObserver in each part of the app.
    * Multiple instances of FeatureObserver are needed by tests only.
    */
-  public static singleton() {
+  public static singleton({bufferSize}: {bufferSize?: number}) {
     if (!__SINGLETON) {
-      __SINGLETON = new FeatureObserver();
+      __SINGLETON = new FeatureObserver({bufferSize});
     }
     return __SINGLETON;
   }
 
+  private _bufferSize = 0;
   private FEATURE_FLAGS: Flags = {values: []};
 
+  constructor({bufferSize = FEATURE_FLAG_BUFFER_SIZE}: {bufferSize?: number}) {
+    this._bufferSize = bufferSize;
+  }
+
   /**
    * Return list of recently accessed feature flags.
    */
   public getFeatureFlags() {
     return this.FEATURE_FLAGS;
   }
 
-  public observeFlags({
-    organization,
+  public updateFlagBuffer({
+    flagName,
+    flagResult,
+    flagBuffer,
     bufferSize,
   }: {
     bufferSize: number;
-    organization: Organization;
+    flagBuffer: Flags;
+    flagName: string;
+    flagResult: boolean;
   }) {
-    const FLAGS = this.FEATURE_FLAGS;
+    // Check if the flag is already in the buffer
+    const index = flagBuffer.values.findIndex(f => f.flag === flagName);
+
+    // The flag is already in the buffer
+    if (index !== -1) {
+      flagBuffer.values.splice(index, 1);
+    }
+
+    // If at capacity, we need to remove the earliest flag
+    // This will only happen if not a duplicate flag
+    if (flagBuffer.values.length === bufferSize) {
+      flagBuffer.values.shift();
+    }
+
+    // Store the flag and its result in the buffer
+    flagBuffer.values.push({
+      flag: flagName,
+      result: flagResult,
+    });
+  }
 
+  public observeOrganizationFlags({organization}: {organization: Organization}) {
+    const flagBuffer = this.FEATURE_FLAGS;
+    const bufferSize = this._bufferSize;
+    const updateFlagBuffer = this.updateFlagBuffer;
     // Track names of features that are passed into the .includes() function.
     const handler = {
-      apply: function (target, orgFeatures, flagName) {
+      apply: (target: any, orgFeatures: string[], flagName: string[]) => {
         // Evaluate the result of .includes()
         const flagResult = target.apply(orgFeatures, flagName);
 
         // Append `feature.organizations:` in front to match the Sentry options automator format
         const name = 'feature.organizations:' + flagName[0];
 
-        // Check if the flag is already in the buffer
-        const index = FLAGS.values.findIndex(f => f.flag === name);
+        updateFlagBuffer({flagName: name, flagResult, flagBuffer, bufferSize});
```

Comment:
`this.updateFlagBuffer(...)` should work in here because `apply: (..) => {` is using the fat-arrow.
 
same for `this._bufferSize`